tools/linter/adapters/header_only_linter.py

```diff
@@ -0,0 +1,102 @@
+#!/usr/bin/env python3
+"""
+Checks that all symbols in torch/header_only_apis.txt are tested in a .cpp
+test file to ensure header-only-ness. The .cpp test file must be built
+without linking libtorch.
+"""
+
+import argparse
+import json
+from enum import Enum
+from pathlib import Path
+from typing import NamedTuple, Union
+
+
+LINTER_CODE = "HEADER_ONLY_LINTER"
+
+
+class LintSeverity(str, Enum):
+    ERROR = "error"
+    WARNING = "warning"
+    ADVICE = "advice"
+    DISABLED = "disabled"
+
+
+class LintMessage(NamedTuple):
+    path: Union[str, None]
+    line: Union[int, None]
+    char: Union[int, None]
+    code: str
+    severity: LintSeverity
+    name: str
+    original: Union[str, None]
+    replacement: Union[str, None]
+    description: Union[str, None]
+
+
+CPP_TEST_GLOBS = [
+    "test/cpp/aoti_abi_check/*.cpp",
+]
+
+REPO_ROOT = Path(__file__).parents[3]
+
+
+def found_symbol(symbol: str) -> bool:
+    # check that the symbol is in a noncommented out region of the test files
+    for cpp_test_glob in CPP_TEST_GLOBS:
+        for test_file in REPO_ROOT.glob(cpp_test_glob):
+            with open(test_file) as tf:
+                for test_file_line in tf:
+                    if not test_file_line.startswith("//") and symbol in test_file_line:
+                        return True
+    return False
+
+
+def check_file(filename: str) -> list[LintMessage]:
+    lint_messages = []
+
+    with open(filename) as f:
+        for idx, line in enumerate(f):
+            # commented out lines should be skipped
+            if line.startswith("#"):
```

Comment:
again ignore leading whitespace? in this case probably just hoist the line.strip() call before this check